### DISCUSSION TITLE
How to Add a Point based on a surface with a PropertySet?

### DISCUSSION THREAD

**Discussion by nonLinearStiffness** [IfcOpenShell/IfcOpenShell#4424] [Q&A] [answered]
Hello,

I'm trying to add, to an existing IFC model, a point that is based on another object (a column, a wall or a floor), and add to that point a set of data (int, doubles, strings, links).

Can someone point me in the right direction, and tell me if it's even possible to accomplish this task?

Thank you very much for your help.

**Comment by aothms** [comment, marked as answer]
> a point that is based on another object

Does the relationship to that object need to be explicitly maintained in the file?

What you describe does not seem very much aligned to how an IFC file is typically structured. We distinguish products and representations, and these can have property sets (user-definable key-value pairs).

What you describe could be a new product (IfcAnnotation) with a placement relative to the column, wall or floor and an IfcShapeRepresentation with a single IfcCartesianPoint. For example. But whether that aligns to your intention is hard to verify with the limited information you provide.

**Reply by nonLinearStiffness** [reply]
The idea is to add inspection data to a structural element in IFC. I don't know if there are any specific ways to do this already?

What I want to do is very similar to a window, but represented with a point.

**Reply by aothms** [reply]
In that case I don't think my approach discussed above is a terrible first attempt.

**Reply by nonLinearStiffness** [reply]
Thank you, I'll try that. Is it possible to make that IfcAnnotation visible in the 3D model?

**Comment by aothms** [comment]
In principle yes. Representations of just a single point are probably not
supported in every viewer, you might want to replace that by a small
sphere.

Sent from a mobile device, excuse my brevity. Kind regards, Thomas

Op do 14 mrt 2024 15:43 schreef nonLinearStifness ***@***.***

**Comment by nonLinearStiffness** [comment]
I just found that there is an actual pSet to deal with Risk Analysis, probably will be using this.
https://standards.buildingsmart.org/IFC/RELEASE/IFC4_3/HTML/lexical/Pset_Risk.htm

**Reply by aothms** [reply]
Yes, and you can see that this can be made applicable to nearly every type of entity in a model (IfcGroup IfcProcess IfcProduct IfcTypeProcess IfcTypeProduct) including something a bit "unsemantic" as an IfcAnnotation.

**Reply by nonLinearStiffness** [reply]
It's good that IFC has that type of flexibility.

What's the way to load these [defaul pSets](https://standards.buildingsmart.org/IFC/DEV/IFC4_2/FINAL/HTML/schema/ifcsharedfacilitieselements/pset/pset_risk.htm)? How can I do it?

**Reply by aothms** [reply]
What do you mean:

> to load

**Reply by nonLinearStiffness** [reply]
Imagine that I want to add that pSet that I've linked above to my annotation. I know how to create a new pSet and add it to the annotation, but is there a way to load the pSet that is already defined in the IFC Standard, or do I need to create a new one based on that information?

What I'm asking is if there is any library that already has these default pSets created?

**Reply by aothms** [reply]
There is a high level API that uses the pset templates specificied in the standard to populate the right data fields https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcopenshell-python/ifcopenshell/api/pset/edit_pset.py But you can also just recreate them manually using the data online.
